File: src/components/tokens.ts
import { existsSync, readFileSync, statSync } from 'node:fs';
import type {
  ComponentConfig,
  ExtendedRenderContext,
  RenderContext,
  TokensComponentConfig,
  TranscriptEntry,
} from '../config/schema.js';
import {
  type AdvancedProgressOptions,
  FINE_PROGRESS_CHARS,
  generateAdvancedProgressBar,
  getRainbowGradientColor,
} from '../utils/index.js';
import { BaseComponent, type ComponentFactory } from './base.js';

/**
 * Token使用信息接口 | Token usage info interface
 */
interface TokenUsageInfo {
  contextUsedTokens: number;
  contextWindow: number;
  usagePercentage: number;
  progressBar?: string;
  warning?: boolean;
  critical?: boolean;
}

/**
 * Tokens组件 | Tokens component
 * 显示当前上下文Token使用情况 | Display current context token usage
 */
export class TokensComponent extends BaseComponent {
  private tokensConfig: TokensComponentConfig;
  private cachedTranscriptData: TokenUsageInfo | null = null;
  private lastTranscriptMtime: number | null = null;

  constructor(name: string, config: TokensComponentConfig) {
    super(name, config);
    this.tokensConfig = config;
  }

  protected renderContent(context: RenderContext): string | null {
    const { inputData } = context;

    // 检查是否有Mock数据 | Check for mock data
    const mockData = (inputData as Record<string, unknown>).__mock__;
    if (mockData && typeof (mockData as Record<string, unknown>).tokenUsage === 'number') {
      return this.renderMockTokenData(
        (mockData as Record<string, unknown>).tokenUsage as number,
        (mockData as Record<string, unknown>).status as string,
        context
      );
    }

    if (!inputData.transcriptPath) {
      return this.renderNoTranscript(context);
    }

    const tokenUsage = this.parseTranscriptFile(inputData.transcriptPath, context);
    if (!tokenUsage) {
      return this.renderNoTranscript(context);
    }

    return this.formatTokenDisplay(tokenUsage);
  }

  /**
   * 渲染Mock数据 | Render mock token data
   */
  private renderMockTokenData(
    tokenUsagePercent: number,
    _status?: string,
    context?: RenderContext
  ): string | null {
    const contextWindow = this.getContextWindow();
    const contextUsedTokens = Math.floor((tokenUsagePercent / 100) * contextWindow);

    const tokenUsage: TokenUsageInfo = {
      contextUsedTokens,
      contextWindow,
      usagePercentage: tokenUsagePercent,
      warning: tokenUsagePercent > (this.tokensConfig.thresholds?.warning || 60),
      critical: tokenUsagePercent > (this.tokensConfig.thresholds?.critical || 95),
    };

    // 生成进度条 | Generate progress bar
    if (this.tokensConfig.show_progress_bar && context) {
      tokenUsage.progressBar = this.generateProgressBar(tokenUsagePercent, context);
    }

    return this.formatTokenDisplay(tokenUsage);
  }

  /**
   * 渲染无transcript文件时的显示 | Render display when no transcript file
   */
  private renderNoTranscript(context: RenderContext): string | null {
    const contextWindow = this.getContextWindow();
    const tokenUsage: TokenUsageInfo = {
      contextUsedTokens: 0,
      contextWindow,
      usagePercentage: 0,
      warning: false,
      critical: false,
    };

    // 生成空进度条 | Generate empty progress bar
    if (this.tokensConfig.show_progress_bar) {
      tokenUsage.progressBar = this.generateProgressBar(0, context);
    }

    return this.formatTokenDisplay(tokenUsage);
  }

  /**
   * 解析transcript文件 | Parse transcript file
   * 简化版：检测压缩并调整计算区间 | Simplified: detect compression and adjust calculation range
   */
  private parseTranscriptFile(
    transcriptPath: string,
    context: RenderContext
  ): TokenUsageInfo | null {
    // 检查文件存在性 | Check file existence
    let fileExists = false;
    try {
      fileExists = existsSync(transcriptPath) && statSync(transcriptPath).isFile();
    } catch (_error) {
      return null;
    }

    if (!fileExists) {
      return {
        contextUsedTokens: 0,
        contextWindow: this.getContextWindow(),
        usagePercentage: 0,
      };
    }

    try {
      const stat = statSync(transcriptPath);
      const currentMtime = stat.mtime.getTime();

      // 检查缓存 | Check cache
      const cacheEnabled = context.config.advanced?.cache_enabled ?? true;
      if (cacheEnabled && this.cachedTranscriptData && this.lastTranscriptMtime === currentMtime) {
        return this.cachedTranscriptData;
      }

      const transcript = readFileSync(transcriptPath, 'utf8');
      const lines = transcript.trim().split('\n');

      // *** 简单压缩检测：检查第一行是否包含 "type":"summary" ***
      // *** Simple compression detection: check if first line contains "type":"summary" ***
      let startLine = 0;
      if (lines.length > 0 && lines[0]?.includes('"type":"summary"')) {
        startLine = 1; // 从第二行开始计算
        if (context.config.debug) {
          console.error('检测到会话压缩，从第', startLine + 1, '行开始计算token');
        }
      }

      let contextUsedTokens = 0;

      // 在指定范围内查找最大usage值 | Find maximum usage value in specified range
      for (let i = lines.length - 1; i >= startLine; i--) {
        const line = lines[i]?.trim();
        if (!line) continue;

        try {
          const entry = JSON.parse(line) as TranscriptEntry;

          // 查找包含usage信息的assistant消息 | Find assistant message with usage info
          if (entry.type === 'assistant' && entry.message && 'usage' in entry.message) {
            const usage = entry.message.usage;

            if (usage && typeof usage === 'object') {
              const inputTokens = Number(usage.input_tokens) || 0;
              const outputTokens = Number(usage.output_tokens) || 0;
              const cacheCreationTokens = Number(usage.cache_creation_input_tokens) || 0;
              const cacheReadTokens = Number(usage.cache_read_input_tokens) || 0;

              const currentUsage =
                inputTokens + outputTokens + cacheCreationTokens + cacheReadTokens;

              if (currentUsage > 0) {
                contextUsedTokens = Math.max(contextUsedTokens, currentUsage);
              }
            }
          }
        } catch (_parseError) {}
      }

      const contextWindow = this.getContextWindow();
      const usagePercentage = (contextUsedTokens / contextWindow) * 100;

      const result: TokenUsageInfo = {
        contextUsedTokens,
        contextWindow,
        usagePercentage,
        warning: usagePercentage > (this.tokensConfig.thresholds?.warning || 60),
        critical: usagePercentage > (this.tokensConfig.thresholds?.critical || 95),
      };

      // 生成进度条 | Generate progress bar
      if (this.tokensConfig.show_progress_bar) {
        result.progressBar = this.generateProgressBar(usagePercentage, context);
      }

      // 缓存结果 | Cache result
      if (cacheEnabled) {
        this.cachedTranscriptData = result;
        this.lastTranscriptMtime = currentMtime;
      }

      return result;
    } catch (error) {
      console.error('Error parsing transcript file:', error);
      return null;
    }
  }

  /**
   * 获取上下文窗口大小 | Get context window size
   */
  private getContextWindow(): number {
    // 从 context_windows 映射中获取，支持模型特定配置
    const contextWindows = this.tokensConfig.context_windows || { default: 200000 };
    return contextWindows.default || 200000;
  }

  /**
   * 生成进度条 | Generate progress bar
   * 支持渐变颜色和精细进度条 | Support gradient colors and fine progress bars
   */
  private generateProgressBar(usagePercentage: number, context: RenderContext): string {
    const width = this.tokensConfig.progress_width || 10;

    // 检查主题配置中的特性启用状态 | Check feature enablement in theme configuration
    const themeConfig = context.config.themes;
    const currentTheme = context.config.theme;

    // 对于powerline和capsule主题，默认启用渐变 | Enable gradient by default for powerline and capsule themes
    const isThemeWithGradient = currentTheme === 'powerline' || currentTheme === 'capsule';
    const enableGradient =
      this.tokensConfig.show_gradient ||
      themeConfig?.[currentTheme]?.enable_gradient ||
      isThemeWithGradient; // 新增：主题默认启用

    const enableFineProgress = themeConfig?.[currentTheme]?.fine_progress || false;

    // 调试信息 | Debug info
    if (context.config.debug) {
      console.error('=== 渐变调试信息 ===');
      console.error('show_gradient:', this.tokensConfig.show_gradient);
      console.error('currentTheme:', currentTheme);
      console.error('theme_enable_gradient:', themeConfig?.[currentTheme]?.enable_gradient);
      console.error('isThemeWithGradient:', isThemeWithGradient);
      console.error('enableGradient:', enableGradient);
      console.error('capabilities.colors:', context.capabilities.colors);
      console.error('=====================');
    }

    // 使用配置的进度条字符 | Use configured progress bar characters
    const filledChar = this.tokensConfig.progress_bar_chars?.filled || '█';
    const emptyChar = this.tokensConfig.progress_bar_chars?.empty || '░';
    const backupChar = this.tokensConfig.progress_bar_chars?.backup || '▓';

    // 智能选择渐变模式 | Intelligently select gradient mode
    const useRainbowGradient = enableGradient;
    const useFineGradient = enableFineProgress && context.capabilities.nerdFont && enableGradient;

    // 准备高级进度条选项 | Prepare advanced progress bar options
    const options: AdvancedProgressOptions = {
      length: width,
      fillChar: filledChar,
      emptyChar: emptyChar,
      backupChar: backupChar,
      backupThreshold: 85,
      enableGradient: useRainbowGradient,
      enableFineProgress: useFineGradient,
      fineChars: FINE_PROGRESS_CHARS,
      colorMapper: getRainbowGradientColor, // 使用新的彩虹渐变算法
    };

    // 调试信息：渐变模式 | Debug info: gradient mode
    if (context.config.debug) {
      console.error('=== 渐变模式调试 ===');
      console.error('useRainbowGradient:', useRainbowGradient);
      console.error('useFineGradient:', useFineGradient);
      console.error('capabilities.nerdFont:', context.capabilities.nerdFont);
      console.error('enableFineProgress:', enableFineProgress);
      console.error('====================');
    }

    // 生成高级进度条 | Generate advanced progress bar
    const result = generateAdvancedProgressBar(usagePercentage, options);

    // 如果启用了渐变并且返回了segments，进行彩色渲染
    // If gradient is enabled and segments are returned, perform colored rendering
    if (enableGradient && result.segments && context.capabilities.colors) {
      if (context.config.debug) {
        console.error('=== 彩色渲染调试 ===');
        console.error('result.segments 存在:', !!result.segments);
        console.error('segments数量:', result.segments?.length);
        console.error('前3个segments:', result.segments?.slice(0, 3));
        console.error('==================');
      }
      return this.renderColoredProgressBar(result.segments, context);
    }

    if (context.config.debug) {
      console.error('=== 未渲染彩色原因 ===');
      console.error('enableGradient:', enableGradient);
      console.error('result.segments存在:', !!result.segments);
      console.error('context.capabilities.colors:', context.capabilities.colors);
      console.error('========================');
    }

    return result.bar;
  }

  /**
   * 渲染彩色进度条 | Render colored progress bar
   * 将分段数据渲染为彩色进度条，支持ANSI RGB代码
   */
  private renderColoredProgressBar(
    segments: Array<{ char: string; color: string }>,
    context: RenderContext
  ): string {
    // 检查是否为 ExtendedRenderContext 并且有 renderer
    const extendedContext = context as ExtendedRenderContext;

    const reset = '\x1b[0m'; // ANSI reset code

    const result = segments
      .map((segment) => {
        // 如果color以\x1b开头，说明是ANSI代码，直接使用
        if (segment.color.startsWith('\x1b')) {
          return `${segment.color}${segment.char}${reset}`;
        }
        // 否则使用渲染器的colorize方法（颜色名称）
        if (extendedContext.renderer && context.capabilities.colors) {
          return extendedContext.renderer!.colorize(segment.char, segment.color);
        }
        // 如果都不行，返回原始字符
        return segment.char;
      })
      .join('');

    if (context.config.debug) {
      console.error('=== 渲染结果调试 ===');
      console.error('输出长度:', result.length);
      console.error('包含ANSI代码:', result.includes('\x1b'));
      console.error('前50个字符:', result.substring(0, 50));
      console.error('===================');
    }

    return result;
  }

  /**
   * 格式化Token显示 | Format token display
   */
  private formatTokenDisplay(tokenUsage: TokenUsageInfo): string {
    const { contextUsedTokens, contextWindow, usagePercentage, progressBar, warning, critical } =
      tokenUsage;

    const icon = this.renderIcon();

    // 确定颜色 | Determine color
    let colorName = 'yellow'; // 默认颜色
    if (this.tokensConfig.colors?.safe) {
      colorName = this.tokensConfig.colors.safe;
    }
    if (critical && this.tokensConfig.colors?.danger) {
      colorName = this.tokensConfig.colors.danger;
    } else if (warning && this.tokensConfig.colors?.warning) {
      colorName = this.tokensConfig.colors.warning;
    }

    // 按原版格式：[进度条] 百分比 (具体数值) | Format like original: [progressbar] percentage (specific numbers)
    let displayText = '';

    // 1. 添加进度条 (带方括号) | Add progress bar (with brackets)
    if (progressBar) {
      displayText += `[${progressBar}] `;
    }

    // 2. 添加百分比 | Add percentage
    if (this.tokensConfig.show_percentage) {
      displayText += `${usagePercentage.toFixed(1)}% `;
    }

    // 3. 添加具体数值 (更精确的k显示) | Add specific numbers (more precise k display)
    const usedDisplay = this.tokensConfig.show_raw_numbers
      ? contextUsedTokens.toString()
      : `${(contextUsedTokens / 1000).toFixed(1)}k`;

    const totalDisplay = this.tokensConfig.show_raw_numbers
      ? contextWindow.toString()
      : `${(contextWindow / 1000).toFixed(0)}k`;

    displayText += `(${usedDisplay}/${totalDisplay})`;

    // 添加状态指示器 - 使用三级图标系统 | Add status indicators - using three-level icon system
    const statusIcon = this.getStatusIcon(critical ?? false, warning ?? false);
    if (statusIcon) {
      displayText += ` ${statusIcon}`;
    }

    return this.formatOutput(icon, displayText, colorName);
  }

  /**
   * 获取状态图标 - 三级图标选择逻辑 | Get status icon - three-level icon selection logic
   * 支持强制参数和优先级：nerd_icon → emoji_icon → text_icon
   */
  private getStatusIcon(critical: boolean, warning: boolean): string {
    if (!critical && !warning) return '';

    const statusType = critical ? 'critical' : 'backup';
    const statusIcons = this.tokensConfig.status_icons;

    if (!statusIcons) {
      // 向后兼容：使用硬编码默认值 | Backward compatibility: use hardcoded defaults
      return critical ? '🔥' : '⚡';
    }

    // 检查是否有强制图标设置（通过renderContext获取配置）
    const context = this.renderContext as ExtendedRenderContext;
    const forceEmoji = context?.config?.terminal?.force_emoji === true;
    const forceNerdFont = context?.config?.terminal?.force_nerd_font === true;
    const forceText = context?.config?.terminal?.force_text === true;

    // 1. 如果强制文本模式
    if (forceText && statusIcons.text?.[statusType]) {
      return statusIcons.text[statusType];
    }

    // 2. 如果强制启用Nerd Font
    if (forceNerdFont && statusIcons.nerd?.[statusType] !== undefined) {
      return statusIcons.nerd[statusType];
    }

    // 3. 如果强制启用emoji
    if (forceEmoji && statusIcons.emoji?.[statusType]) {
      return statusIcons.emoji[statusType];
    }

    // 4. 自动检测模式：优先使用Nerd Font图标（如果支持）
    if (this.capabilities.nerdFont && statusIcons.nerd?.[statusType] !== undefined) {
      return statusIcons.nerd[statusType];
    }

    // 5. 自动检测模式：其次使用Emoji图标（如果支持）
    if (this.capabilities.emoji && statusIcons.emoji?.[statusType]) {
      return statusIcons.emoji[statusType];
    }

    // 6. 最后回退到文本图标
    if (statusIcons.text?.[statusType]) {
      return statusIcons.text[statusType];
    }

    // 最后的回退：硬编码默认值 | Final fallback: hardcoded defaults
    return critical ? '[X]' : '[!]';
  }
}

/**
 * Tokens组件工厂 | Tokens component factory
 */
export class TokensComponentFactory implements ComponentFactory {
  createComponent(name: string, config: ComponentConfig): TokensComponent {
    return new TokensComponent(name, config as TokensComponentConfig);
  }

  getSupportedTypes(): string[] {
    return ['tokens'];
  }
}

File: src/components/tokens.rs
//! Tokens component implementation
//!
//! Displays token usage information with cached transcript statistics and adaptive progress bars.

use std::fmt::Write;

use async_trait::async_trait;

use super::base::{Component, ComponentFactory, ComponentOutput, RenderContext};
use crate::config::{BaseComponentConfig, Config, TokensComponentConfig};
use crate::storage;

#[derive(Clone, Debug)]
struct TokenUsageInfo {
    used: u64,
    total: u64,
}

/// Tokens component
pub struct TokensComponent {
    config: TokensComponentConfig,
}

impl TokensComponent {
    #[must_use]
    pub const fn new(config: TokensComponentConfig) -> Self {
        Self { config }
    }

    async fn fetch_usage_from_cache(&self, ctx: &RenderContext) -> Option<TokenUsageInfo> {
        if let Some(mock_tokens) = ctx
            .input
            .extra
            .get("__mock__")
            .and_then(|mock| mock.get("tokensUsage"))
        {
            let used = mock_tokens
                .get("context_used")
                .and_then(serde_json::Value::as_u64)
                .unwrap_or(0);
            if used == 0 && !self.config.show_zero {
                return None;
            }
            let window = mock_tokens
                .get("context_window")
                .and_then(serde_json::Value::as_u64)
                .unwrap_or_else(|| self.context_window_for_model(ctx));
            return Some(TokenUsageInfo {
                used,
                total: window,
            });
        }

        if let Some(session_id) = ctx.input.session_id.as_deref() {
            if let Ok(Some(tokens)) = storage::get_session_tokens(session_id).await {
                let used = tokens.context_used;
                if used == 0 && !self.config.show_zero {
                    return None;
                }
                let window = self.context_window_for_model(ctx);
                return Some(TokenUsageInfo {
                    used,
                    total: window,
                });
            }
        }
        if self.config.show_zero {
            let window = self.context_window_for_model(ctx);
            return Some(TokenUsageInfo {
                used: 0,
                total: window,
            });
        }
        None
    }

    fn context_window_for_model(&self, ctx: &RenderContext) -> u64 {
        let default_window = self
            .config
            .context_windows
            .get("default")
            .copied()
            .unwrap_or(200_000);

        let Some(model) = ctx.input.model.as_ref() else {
            return default_window;
        };

        if let Some(id) = model.id.as_ref() {
            if let Some(value) = self.config.context_windows.get(id) {
                return *value;
            }
        }

        default_window
    }

    fn build_progress_bar(&self, ctx: &RenderContext, percentage: f64) -> Option<String> {
        if !self.config.show_progress_bar {
            return None;
        }

        let width = self.config.progress_width.max(1) as usize;
        let width_f64 = to_f64(width);
        let filled_len = clamp_round_to_usize((percentage / 100.0) * width_f64, width);
        let capped_filled = filled_len.min(width);

        let gradient_enabled = self.config.show_gradient
            || matches!(ctx.config.theme.as_str(), "powerline" | "capsule");
        let supports_colors = ctx.terminal.supports_colors;

        let filled_char = self
            .config
            .progress_bar_chars
            .filled
            .chars()
            .next()
            .unwrap_or('█');
        let empty_char = self
            .config
            .progress_bar_chars
            .empty
            .chars()
            .next()
            .unwrap_or('░');
        let backup_char = self
            .config
            .progress_bar_chars
            .backup
            .chars()
            .next()
            .unwrap_or('▓');

        let mut bar = String::with_capacity(width * 16);
        let mut color_active = false;

        for idx in 0..width {
            if idx < capped_filled {
                let gradient_percentage = if capped_filled == 0 {
                    0.0
                } else {
                    let idx_f64 = to_f64(idx);
                    let capped_filled_f64 = to_f64(capped_filled);

                    ((idx_f64 + 0.5) / capped_filled_f64) * percentage
                }
                .clamp(0.0, 100.0);
                let is_backup = gradient_percentage >= self.config.thresholds.backup;
                let symbol = if is_backup { backup_char } else { filled_char };

                if gradient_enabled && supports_colors {
                    let (r, g, b) = rainbow_gradient_color(gradient_percentage);
                    let _ = write!(bar, "\x1b[38;2;{r};{g};{b}m{symbol}");
                    color_active = true;
                } else {
                    bar.push(symbol);
                }
            } else if gradient_enabled && supports_colors {
                bar.push_str("\x1b[38;2;120;120;120m");
                bar.push(empty_char);
                color_active = true;
            } else {
                bar.push(empty_char);
            }
        }

        if color_active {
            bar.push_str("\x1b[0m");
        }

        Some(bar)
    }

    fn select_status_icon(&self, ctx: &RenderContext, percentage: f64) -> Option<String> {
        let thresholds = &self.config.thresholds;
        let status = if percentage >= thresholds.critical {
            TokenStatusKind::Critical
        } else if percentage >= thresholds.backup {
            TokenStatusKind::Backup
        } else {
            return None;
        };

        let icons = &self.config.status_icons;
        let terminal_cfg = &ctx.config.terminal;
        let terminal = &ctx.terminal;
        let style = &ctx.config.style;

        if terminal_cfg.force_text {
            return icon_for_kind(&icons.text, status).map(std::string::ToString::to_string);
        }
        if terminal_cfg.force_nerd_font {
            if let Some(icon) = icon_for_kind(&icons.nerd, status) {
                return Some(icon.to_string());
            }
        }
        if terminal_cfg.force_emoji {
            if let Some(icon) = icon_for_kind(&icons.emoji, status) {
                return Some(icon.to_string());
            }
        }

        if terminal.supports_nerd_font
            && style
                .enable_nerd_font
                .is_enabled(terminal.supports_nerd_font)
        {
            if let Some(icon) = icon_for_kind(&icons.nerd, status) {
                return Some(icon.to_string());
            }
        }

        if terminal.supports_emoji && style.enable_emoji.is_enabled(terminal.supports_emoji) {
            if let Some(icon) = icon_for_kind(&icons.emoji, status) {
                return Some(icon.to_string());
            }
        }

        icon_for_kind(&icons.text, status).map(std::string::ToString::to_string)
    }

    fn select_color(&self, percentage: f64) -> String {
        let thresholds = &self.config.thresholds;

        if percentage >= thresholds.danger {
            self.config.colors.danger.clone()
        } else if percentage >= thresholds.warning {
            self.config.colors.warning.clone()
        } else {
            self.config.colors.safe.clone()
        }
    }

    fn format_usage(&self, info: &TokenUsageInfo) -> String {
        if self.config.show_raw_numbers {
            format!("({}/{})", info.used, info.total)
        } else {
            let used_k = to_f64(info.used) / 1_000.0;
            let total_k = to_f64(info.total) / 1_000.0;
            format!("({used_k:.1}k/{total_k:.0}k)")
        }
    }
}

#[async_trait]
impl Component for TokensComponent {
    fn name(&self) -> &'static str {
        "tokens"
    }

    fn is_enabled(&self, _ctx: &RenderContext) -> bool {
        self.config.base.enabled
    }

    async fn render(&self, ctx: &RenderContext) -> ComponentOutput {
        if !self.is_enabled(ctx) {
            return ComponentOutput::hidden();
        }

        let Some(usage) = self.fetch_usage_from_cache(ctx).await else {
            return ComponentOutput::hidden();
        };

        let total = usage.total.max(1);
        let percentage = (to_f64(usage.used) / to_f64(total)) * 100.0;
        let clamped_percentage = percentage.clamp(0.0, 999.9);

        let mut parts = Vec::new();

        if let Some(bar) = self.build_progress_bar(ctx, clamped_percentage) {
            parts.push(format!("[{bar}]"));
        }

        if self.config.show_percentage {
            parts.push(format!("{clamped_percentage:.1}%"));
        }

        parts.push(self.format_usage(&usage));

        if let Some(status_icon) = self.select_status_icon(ctx, clamped_percentage) {
            parts.push(status_icon);
        }

        let text = parts.join(" ");
        let color = self.select_color(clamped_percentage);
        let icon = self.select_icon(ctx);

        ComponentOutput::new(text)
            .with_icon(icon.unwrap_or_default())
            .with_icon_color(color.clone())
            .with_text_color(color)
    }

    fn base_config(&self, _ctx: &RenderContext) -> Option<&BaseComponentConfig> {
        Some(&self.config.base)
    }
}

fn icon_for_kind(set: &crate::config::TokenIconSetConfig, kind: TokenStatusKind) -> Option<&str> {
    match kind {
        TokenStatusKind::Backup => (!set.backup.is_empty()).then_some(set.backup.as_str()),
        TokenStatusKind::Critical => (!set.critical.is_empty()).then_some(set.critical.as_str()),
    }
}

#[derive(Clone, Copy)]
enum TokenStatusKind {
    Backup,
    Critical,
}

fn rainbow_gradient_color(percentage: f64) -> (u8, u8, u8) {
    let p = percentage.clamp(0.0, 100.0);

    let soft_green = (80.0, 200.0, 80.0);
    let soft_yellow_green = (150.0, 200.0, 60.0);
    let soft_yellow = (200.0, 200.0, 80.0);
    let soft_orange = (220.0, 160.0, 60.0);
    let soft_red = (200.0, 100.0, 80.0);

    let lerp = |start: (f64, f64, f64), end: (f64, f64, f64), t: f64| {
        let clamp_t = t.clamp(0.0, 1.0);
        (
            (end.0 - start.0).mul_add(clamp_t, start.0),
            (end.1 - start.1).mul_add(clamp_t, start.1),
            (end.2 - start.2).mul_add(clamp_t, start.2),
        )
    };

    let (r, g, b) = if p <= 25.0 {
        lerp(soft_green, soft_yellow_green, p / 25.0)
    } else if p <= 50.0 {
        lerp(soft_yellow_green, soft_yellow, (p - 25.0) / 25.0)
    } else if p <= 75.0 {
        lerp(soft_yellow, soft_orange, (p - 50.0) / 25.0)
    } else {
        lerp(soft_orange, soft_red, (p - 75.0) / 25.0)
    };

    let convert = |value: f64| -> u8 {
        #[allow(clippy::cast_possible_truncation, clippy::cast_sign_loss)]
        {
            value.clamp(0.0, 255.0).round() as u8
        }
    };

    (convert(r), convert(g), convert(b))
}

fn clamp_round_to_usize(value: f64, max: usize) -> usize {
    let max_f64 = to_f64(max);
    let clamped = value.clamp(0.0, max_f64);

    #[allow(clippy::cast_possible_truncation, clippy::cast_sign_loss)]
    let rounded = clamped.round() as usize;

    rounded.min(max)
}

fn to_f64<T: IntoF64>(value: T) -> f64 {
    value.into_f64()
}

trait IntoF64 {
    fn into_f64(self) -> f64;
}

impl IntoF64 for usize {
    fn into_f64(self) -> f64 {
        #[allow(clippy::cast_precision_loss)]
        {
            self as f64
        }
    }
}

impl IntoF64 for u64 {
    fn into_f64(self) -> f64 {
        #[allow(clippy::cast_precision_loss)]
        {
            self as f64
        }
    }
}

/// Factory for creating Tokens components
pub struct TokensComponentFactory;

impl ComponentFactory for TokensComponentFactory {
    fn create(&self, config: &Config) -> Box<dyn Component> {
        Box::new(TokensComponent::new(config.components.tokens.clone()))
    }

    fn name(&self) -> &'static str {
        "tokens"
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::components::TerminalCapabilities;
    use crate::config::AutoDetect;
    use crate::core::InputData;
    use serde_json::json;
    use std::sync::Arc;

    #[allow(clippy::field_reassign_with_default)]
    fn build_tokens_config(
        configure: impl FnOnce(&mut TokensComponentConfig),
    ) -> TokensComponentConfig {
        let mut config = TokensComponentConfig::default();
        configure(&mut config);
        config
    }

    #[allow(clippy::field_reassign_with_default)]
    fn build_input(configure: impl FnOnce(&mut InputData)) -> InputData {
        let mut input = InputData::default();
        configure(&mut input);
        input
    }

    fn create_test_context_with_tokens(tokens: i64) -> RenderContext {
        let used = u64::try_from(tokens).unwrap_or(0);

        let input = build_input(|input| {
            input.session_id = Some("mock-session".to_string());
            input.extra = json!({
                "__mock__": {
                    "tokensUsage": {
                        "context_used": used
                    }
                }
            });
        });

        RenderContext {
            input: Arc::new(input),
            config: Arc::new(Config::default()),
            terminal: TerminalCapabilities::default(),
        }
    }

    #[tokio::test]
    async fn test_tokens_contains_percentage() {
        let component = TokensComponent::new(TokensComponentConfig::default());
        let ctx = create_test_context_with_tokens(1_000);

        let output = component.render(&ctx).await;
        assert!(output.visible);
        assert!(output.text.contains('%'));
    }

    #[tokio::test]
    async fn test_tokens_raw_numbers_format() {
        let config = build_tokens_config(|config| {
            config.show_percentage = false;
            config.show_progress_bar = false;
            config.show_raw_numbers = true;
        });

        let component = TokensComponent::new(config);
        let ctx = create_test_context_with_tokens(1_500);

        let output = component.render(&ctx).await;
        assert!(output.visible);
        assert!(output.text.contains("(1500/200000)"));
    }

    #[tokio::test]
    async fn test_tokens_progress_bar_enabled() {
        let config = build_tokens_config(|config| {
            config.show_progress_bar = true;
            config.show_percentage = false;
            config.show_raw_numbers = false;
        });

        let component = TokensComponent::new(config);
        let ctx = create_test_context_with_tokens(50_000);

        let output = component.render(&ctx).await;
        assert!(output.visible);
        assert!(output.text.contains('['));
    }

    #[tokio::test]
    async fn test_tokens_progress_bar_gradient() {
        let config = build_tokens_config(|config| {
            config.show_progress_bar = true;
            config.show_percentage = false;
            config.show_raw_numbers = false;
            config.show_gradient = true;
            config.progress_width = 6;
        });

        let component = TokensComponent::new(config);
        let mut ctx = create_test_context_with_tokens(100_000);
        let config = Arc::make_mut(&mut ctx.config);
        config.theme = "classic".to_string();
        config.style.enable_colors = AutoDetect::Bool(true);
        let mut terminal = ctx.terminal.clone();
        terminal.supports_colors = true;
        let ctx = RenderContext { terminal, ..ctx };

        let output = component.render(&ctx).await;
        assert!(output.visible);
        assert!(output.text.contains("\x1b[38;2"));
    }

    #[tokio::test]
    async fn test_tokens_zero_hidden() {
        let config = build_tokens_config(|config| {
            config.show_zero = false;
        });

        let component = TokensComponent::new(config);
        let ctx = create_test_context_with_tokens(0);

        let output = component.render(&ctx).await;
        assert!(!output.visible);
    }

    #[tokio::test]
    async fn test_tokens_zero_shown() {
        let config = build_tokens_config(|config| {
            config.show_zero = true;
        });

        let component = TokensComponent::new(config);
        let ctx = create_test_context_with_tokens(0);

        let output = component.render(&ctx).await;
        assert!(output.visible);
    }

    #[tokio::test]
    async fn test_tokens_disabled() {
        let config = build_tokens_config(|config| {
            config.base.enabled = false;
        });

        let component = TokensComponent::new(config);
        let ctx = create_test_context_with_tokens(1000);

        let output = component.render(&ctx).await;
        assert!(!output.visible);
    }

    #[tokio::test]
    async fn test_tokens_mock_context_window_override() {
        let input = build_input(|input| {
            input.session_id = Some("mock-session".to_string());
            input.extra = json!({
                "__mock__": {
                    "tokensUsage": {
                        "context_used": 20u64,
                        "context_window": 100u64
                    }
                }
            });
        });

        let ctx = RenderContext {
            input: Arc::new(input),
            config: Arc::new(Config::default()),
            terminal: TerminalCapabilities::default(),
        };

        let config = build_tokens_config(|config| {
            config.show_progress_bar = false;
            config.show_percentage = false;
            config.show_raw_numbers = true;
        });

        let component = TokensComponent::new(config);
        let output = component.render(&ctx).await;

        assert!(output.visible);
        assert!(output.text.contains("(20/100)"));
    }
}
